edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Service/Program.cs

```diff
@@ -122,21 +115,23 @@ static async Task<int> MainAsync(IConfigurationRoot configuration)
             TimeSpan shutdownWaitPeriod = TimeSpan.FromSeconds(configuration.GetValue("ShutdownWaitPeriod", DefaultShutdownWaitPeriod));
             (CancellationTokenSource cts, ManualResetEventSlim completed, Option<object> handler) = ShutdownHandler.Init(shutdownWaitPeriod, logger);
 
-            using (IProtocolHead protocolHead = await GetEdgeHubProtocolHeadAsync(logger, configuration, experimentalFeatures, container, hosting))
+            using (IProtocolHead mqttBrokerProtocolHead = await GetMqttBrokerProtocolHeadAsync(experimentalFeatures, container))
+            using (IProtocolHead edgeHubProtocolHead = await GetEdgeHubProtocolHeadAsync(logger, configuration, experimentalFeatures, container, hosting))
             using (var renewal = new CertificateRenewal(certificates, logger))
             {
                 try
                 {
-                    await protocolHead.StartAsync();
+                    await Task.WhenAll(mqttBrokerProtocolHead.StartAsync(), configDownloadTask);
                     await Task.WhenAny(cts.Token.WhenCanceled(), renewal.Token.WhenCanceled(), configUpdaterStartupFailed.Task);
+                    await edgeHubProtocolHead.StartAsync();
                 }
```

Comment:
Is the order right? The way I read the code is that it starts mqttbroker (or the empty protocol head) and the configDownload (this part is ok)
But then it waits for shutdown (or till the certificate expires or the config download failed) and then starts the rest of the protocol head.

Reply:
Nice catch. updated.